_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+
+The `session.timeout.ms` property specifies the maximum amount of time a consumer in a group can go without sending a _heartbeat_ to the Kafka broker. 
+This heartbeat is an indicator that the consumer is active in the group. 
+If a consumer fails to send a heartbeat within the specified session timeout, Kafka considers it as a failure, and the consumer is marked as inactive.
+A consumer marked as inactive triggers a rebalancing of the partitions for the topic.
+Setting this value too low can result in false-positive outcomes, while setting it too high can lead to delayed recovery from failures.
+
+The `heartbeat.interval.ms` property determines how frequently a consumer sends heartbeats to the Kafka broker. 
+A shorter interval between consecutive heartbeats allows for quicker detection of consumer failures.
+The heartbeat interval must be lower, usually by a third, than the session timeout.
+Decreasing the heartbeat interval reduces the chance of accidental rebalancing, but more frequent heartbeats increases the overhead on broker resources.
+
+**Minimizing the impact of rebalances**
+
+The consumer application uses the `onPartitionsRevoked` method so that pending offsets are committed before the consumer loses ownership of the partition during a rebalance.
+
+To further reduce potential disruptions caused by rebalances, you can use the `group.instance.id` to assign a unique identifier to each consumer instance within the consumer group (as specified by `GROUP_ID` in our consumer).
+This approach minimizes unnecessary partition rebalancing when a consumer rejoins a group after a failure or restart. The consumer continues with the same instance ID and maintains its assignment of topic partitions.
+
+Additionally, adjusting the `max.poll.interval.ms` configuration can prevent rebalances caused by prolonged processing tasks, allowing you to specify the maximum interval between polls for new messages. If lengthy message processing is unavoidable due to factors such as slow external services, consider offloading this processing to a pool of worker threads. 
+Similarly `max.poll.records` limits the number of records returned from the consumer buffer, allowing the consumer application to process fewer messages more efficiently.
+
+## Get the message
+
+In this blog post, we've explored the essentials of creating Kafka consumer applications, covering everything from handling offsets and errors to fine-tuning performance and minimizing the impact of rebalances.
+
+A successful consumer client is all about effectively receiving and processing data from Kafka topics. Just like with producers, the effectiveness of your consumer application depends on your specific use case and requirements, as well as the quality of the data provided by producers.
+
+Kafka's architecture and features make it a powerful platform for end-to-end data streaming. 
```

Comment:
Rather than `end to end`, I wonder if `real time` and/or `scalable` would be better adjectives to describe Kafka.